controllers/etcdcopybackupstask/reconciler_test.go

```diff
@@ -84,36 +84,54 @@ var _ = Describe("EtcdCopyBackupsTaskController", func() {
 		var (
 			ctx        = context.Background()
 			task       *druidv1alpha1.EtcdCopyBackupsTask
-			fakeClient = fakeclient.NewClientBuilder().WithScheme(kubernetes.Scheme).Build()
-			r          = &Reconciler{
-				Client: fakeClient,
-				logger: logr.Discard(),
-			}
+			fakeClient client.WithWatch
+			r          *Reconciler
 		)
 		const (
-			testTaskName  = "test-etcd-backup-copy-task"
+			testTaskName  = "test-etcd-copy-backups-task"
 			testNamespace = "test-ns"
 		)
 
-		Context("delete EtcdCopyBackupsTask object tests when it exists", func() {
+		Context("delete EtcdCopyBackupsTask object when it exists", func() {
 			BeforeEach(func() {
-				task = ensureEtcdCopyBackupsTaskCreation(ctx, testTaskName, testNamespace, fakeClient)
+				task = testutils.CreateEtcdCopyBackupsTask(testTaskName, testNamespace, "aws", false)
+
+				By("Create fake client with task object")
+				fakeClient = fakeclient.NewClientBuilder().
+					WithScheme(kubernetes.Scheme).
+					WithObjects(task).
+					WithStatusSubresource(task).
+					Build()
+
+				By("Ensure that copy backups task is created")
+				Eventually(func() error {
+					return fakeClient.Get(ctx, client.ObjectKeyFromObject(task), task)
+				}).Should(Succeed())
+
+				r = &Reconciler{
+					Client: fakeClient,
+					logger: logr.Discard(),
+				}
 			})
 			AfterEach(func() {
 				ensureEtcdCopyBackupsTaskRemoval(ctx, testTaskName, testNamespace, fakeClient)
 			})
 
-			It("should not delete if there is no deletion timestamp set and no finalizer set", func() {
+			It("should not delete task if there is no deletion timestamp set and no finalizer set", func() {
 				_, err := r.delete(ctx, task)
 				Expect(err).To(BeNil())
 				foundTask := &druidv1alpha1.EtcdCopyBackupsTask{}
 				Expect(fakeClient.Get(ctx, client.ObjectKeyFromObject(task), foundTask)).To(Succeed())
 				Expect(client.ObjectKeyFromObject(foundTask)).To(Equal(client.ObjectKeyFromObject(task)))
 			})
 
-			It("should remove finalizer for task which does not have a corresponding job", func() {
+			It("should remove finalizer and delete task if it does not have a corresponding job", func() {
 				Expect(controllerutils.AddFinalizers(ctx, fakeClient, task, common.FinalizerName)).To(Succeed())
-				Expect(addDeletionTimestampToTask(ctx, task, time.Now(), fakeClient)).To(Succeed())
+				// use fakeClient.Delete() to simply add deletionTimestamp to `task` object,
+				// due to https://github.com/kubernetes-sigs/controller-runtime/pull/2316
+				Expect(fakeClient.Delete(ctx, task)).To(Succeed())
+				// get the updated object after deletionTimestamp has been added by fakeClient.Delete() call
+				Expect(fakeClient.Get(ctx, client.ObjectKeyFromObject(task), task)).To(Succeed())
```

Comment:
Doesn't the `Delete` call already work on a pointer to the object? Is this `Get` call still needed?

Reply:
> Doesn't the Delete call already work on a pointer to the object?

Yes, but within the deletion logic of fake client, it [adds deletion timestamp on a copy of the object](https://github.com/kubernetes-sigs/controller-runtime/blob/5f8d96b043ef8ccadc544535415395be336630e8/pkg/client/fake/client.go#L1031-L1047) (yes, quite weird). So the original object we had passed then differs from the actual object in the cluster, so we are hit with this error:

```
 [FAILED] Expected
      <*fmt.wrapError | 0x14000781aa0>: 
      could not remove finalizer: Operation cannot be fulfilled on etcdcopybackupstasks.druid.gardener.cloud "test-etcd-copy-backups-task": object was modified
      {
          msg: "could not remove finalizer: Operation cannot be fulfilled on etcdcopybackupstasks.druid.gardener.cloud \"test-etcd-copy-backups-task\": object was modified",
          err: <*errors.StatusError | 0x140005ff720>{
              ErrStatus: {
                  TypeMeta: {Kind: "", APIVersion: ""},
                  ListMeta: {
                      SelfLink: "",
                      ResourceVersion: "",
                      Continue: "",
                      RemainingItemCount: nil,
                  },
                  Status: "Failure",
                  Message: "Operation cannot be fulfilled on etcdcopybackupstasks.druid.gardener.cloud \"test-etcd-copy-backups-task\": object was modified",
                  Reason: "Conflict",
                  Details: {
                      Name: "test-etcd-copy-backups-task",
                      Group: "druid.gardener.cloud",
                      Kind: "etcdcopybackupstasks",
                      UID: "",
                      Causes: nil,
                      RetryAfterSeconds: 0,
                  },
                  Code: 409,
              },
          },
      }
  to be nil
  In [It] at: /Users/i349079/go/src/github.com/gardener/etcd-druid/controllers/etcdcopybackupstask/reconciler_test.go:137 @ 01/09/24 12:20:24.181
```

The only way to circumvent this error is to get the object again and then continue.

Reply:
Weird, but anyways, for the tests, we can live with it.